Add sketch compilation tests

bduffany opened the pull request:
This branch (which resolves #3 and #6) introduces tests for the following pages:
- `/libraries`
- `/user/cb_compile_tester`

It tests that every single sketch compiles. For the latter, we keep a log of all successes and failures in the logs directory. The logs have the following format:

```
{
  "url": "http://localhost", // (or http://codebender.cc) 
  "succeeded": ["<sketch_url>", "<sketch_url>", ...],
  "failed": [
    {
      "sketch": "<sketch_url>",
      "exception": "<exception message>"
    },
    ... 
  ]
}
```

For some reason, it appears that the vast majority of `cb_compile_tester`'s sketches are failing to verify. I'm not sure yet whether that's something to do with the script, or whether those sketches actually have errors.

For `/libraries`, I don't actually do any logging, but I have the framework in place so that adding logging for it would be extremely simple.

fpapadopou commented:
@bduffany yes, the format is fine. As for the sketches, they shouldn't all compile successfully, there are still edge cases that make some of the projects fail. 

Please, do not push any more commits to the pull request, because we will have to make some changes to the issue.

bduffany commented:
I just implemented the `--url` functionality. The user can now run `py.test --url http://codebender.cc tests/compile_tester` (for example) to compile all of `/user/cb_compile_tester`'s libraries. However I also added the functionality to conditionally skip tests depending on the site. e.g. `py.test --url http://localhost tests/compile_tester` will result in a skipped test, since `/user/cb_compile_tester` isn't actually a user in the bachelor version.

spiliot commented:
Up to 2fb1e02 :+1:

bduffany commented:
@spiliot git doesn't commit empty directories, and I wanted to make sure the directory itself got committed, since the code relies on that directory to exist in order to generate log files. I'll document that in the `.gitignore` itself

fpapadopou commented:
Good job with the libraries. We'll have to implement the project upload for the cb_compile_tester's projects in order to close the pull request. Keep in mind we also need to test cb_compile_tester's projects and bachelor's library examples separately.

bduffany commented:
@fpapadopou the automated uploading/compiling/deleting for cb_compile_tester in localhost is now done in #7

spiliot commented:
@bduffany what I (and most people) do to commit empty dirs is to add an `empty` file (that is a file called empty with 0 size). Not that you should change this now but you can keep it in mind.

bduffany commented:
I would also like these changes to be in master; I'm not sure quite how that will work in terms of merging since this pull request was to merge into `bduffany-selenium1`.

tsiknas commented:
closing since it is already merged with master
